### ISSUE TITLE
`test_bump_recipe_cli()` for multi-sourced recipes dramatically slows tests down (caused by retrying on mocked HTTP calls?)

### ISSUE THREAD

**Issue by schuylermartin45** [conda-incubator/conda-recipe-manager#265]
As described/discovered/introduced in #260/#264, concurrently acquiring multiple source artifacts causes a significant slow down in the unit tests. The two tests that fall into this category reference:
- `gsm-amzn2-aarch64_version_bump.yaml`
- `libprotobuf_version_bump.yaml`

On occasion, these two tests now take about 30 seconds each. Most of this is mitigated by `xdist` parallelization, but on occassion if both tests are run sequentially, the tests take 60 seconds to complete.

As far as I can currently tell, it seems like `pyfakefs` is the cause of this slow down, but it is unclear why. Disabling `pyfakefs` usage seems to remove the issue. Not all runs suffer from this slow down. Presumably `pyfakefs` is causing some kind of contention that does not result in a complete deadlock.

**UPDATE**
As discussed in the comments below, the reason why the test takes so long is obvious. We need to significantly reduce the retry time limit in the test. What is more interesting/still unknown is _why_ we are sporadically hitting the HTTP retry mechanism on valid mocked HTTP calls.

**Comment by schuylermartin45** [comment]
Today on my local machine I've had the tests take 7+ minutes, which is a dramatic and very annoying change. Still stumped at the root cause.

**Comment by schuylermartin45** [comment]
Found this: https://github.com/pytest-dev/pyfakefs/blob/main/docs/troubleshooting.rst#modules-not-working-with-pyfakefs

**Comment by schuylermartin45** [comment]
Asked for clarification here: https://github.com/pytest-dev/pyfakefs/issues/1102

**Comment by schuylermartin45** [comment]
No worries! And I really appreciate you going out of your way to look into this!

Now it has been a few months since I've thought about this, but _I think_ I left the retry interval as-is to act as a canary. With the exception of the actual retry test, we should never be retrying on mocked HTTP calls.

That being said, I should probably put in the retry mitigation now to save compute and flag it with a comment referencing this ticket/investigation.

I guess I should also rename and update the issue to indicate that we're more curious about _why_ the retries happen. Not why the tests take so long.

**Comment by mrbean-bremen** [comment]
Ah ok, that makes sense! So far I only found that it has probably to do with the fact that the packages in question have more than one URL (and SHA), which seem to be handled in parallel. `pyfakefs` does nothing to be thread-safe (though that being Python, it shouldn't need to, at least in normal Python code), and there seems to be some timing problem here.

**Comment by schuylermartin45** [comment]
I'd have to double check, but _I think_ the parallelism mechanisms I'm using bypass the global lock to perform better with I/O bound ops.

So I think a resource contention issue makes sense.

**Comment by mrbean-bremen** [comment]
> I think the parallelism mechanisms I'm using bypass the global lock to perform better

That could indeed explain it. That is something that I very much doubt pyfakefs could handle, so maybe the current workaround is the best we can do here. Is this the `ThreadPoolExecutor` you are speaking about here?

**Comment by schuylermartin45** [comment]
Yeah I _think_ that avoids the GIL. And now that we're talking about this, I just remembered that I have an outstanding issue to rework the temporary file system to use a context manager (See #299).

So it could be something along the lines of the temp folder is not always being garbage-collected as soon as it could be. I think I had my C++ brain on when I wrote it and assumed it'd fall out of scope, lol.

**Comment by mrbean-bremen** [comment]
Well, a context manager in Python is not so far from a RAII class in C++ :) 

I just checked, I think the `ThreadPoolExecutor` itself will not avoid the GIL, though it is faster if used with tasks which release the GIL - not sure if that happens here.

**Comment by schuylermartin45** [comment]
Ah okay then I think [requests](https://github.com/psf/requests/issues/3474) releases the lock via `sockets` and I'm getting my wires crossed. That makes a bit more sense.

**Comment by mrbean-bremen** [comment]
Ah ok - I thought that requests is patched, but obviously not all of it.

**Comment by mrbean-bremen** [comment]
Ok, I had another look after what I've learned, and I'm pretty sure that the problem is an artifact of the `requests` mock - specifically, that it accesses the same dummy file with write access from different threads.
If I add a lock in the mock call, the problem goes away:
```py
from threading import Lock

_lock = Lock()
...

    match endpoint:
        case endpoint if endpoint in default_artifact_set:
            _lock.acquire()
            try:
                return MockHttpStreamResponse(200, f"archive_files/dummy_project_01.tar.gz")
            finally:
                _lock.release()
```

**Comment by mrbean-bremen** [comment]
Yes, the underlying issue here is probably that `pyfakefs` does not lock files on writing. This is OS-dependent: under Windows, you cannot open a file for writing twice, as it is done here, so the behavior is as expected, but under Linux, as far as I know there is a write lock, but not a lock on the underlying stream - though I'm not sure how much of this behavior is guaranteed, would need to look it up. Anyway, `pyfakefs` is not designed for multi-threading, which is something I'm reluctant to change (it may open a whole new can of worms), but which should at least be better documented.

What I meant above was that your problem only appears only because of the mock behavior, the real code would not behave this way - at least this is what I understood.

Anyway, thanks for working with me here! It is always very helpful for me to understand the problems that appear in real code working with pyfakefs.
